### ISSUE TITLE
Remove `null_count` from ArrayData::try_new()

### ISSUE THREAD

**Issue by alamb** [apache/arrow-rs#911]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Re #817 ; As @jhorstmann  pointed out on #810 at https://github.com/apache/arrow-rs/pull/810/files#r742905920

`ArrayData::try_new()` function includes an optional `null_count` argument and a validity buffer. If these numbers differ,  wrong results can occur (and maybe also undefined behavior) in some later operation. 

Note that as part of #817 we will be validating that the bitmaps are consistent with the declared `null_count`

**Describe the solution you'd like**

Since most callers pass `None` anyway at which point we calculate the number, if we simply removed the option to pass in an inconsistent value we would avoid a class of inconsistencies I would suggest to remove the `null_count` parameter from `try_new` to completely avoid of inconsistencies.

**Describe alternatives you've considered**
N/A 

**Additional context**

**Comment by HaoYang670** [comment]
If we need to implement this, I can have a try!

**Comment by alamb** [comment]
> If we need to implement this, I can have a try!

I think it is still worth trying 👍

**Comment by HaoYang670** [comment]
I find the above bug in the test :
```rust
    #[test]
    #[should_panic(expected = "null_bit_buffer size too small. got 1 needed 2")]
    fn test_bitmap_too_small() {
        let buffer = make_i32_buffer(9);
        let null_bit_buffer = Buffer::from(vec![0b11111111]);

        ArrayData::try_new(
            DataType::Int32,
            9,
            Some(0),
            Some(null_bit_buffer),
            0,
            vec![buffer],
            vec![],
        )
        .unwrap();
    }
```

Here we will not calculate the null count in `new_unchecked` because we have provided a null count `Some(0)`. And when we validate the array data, we will get the error `ArrayError(null_bit_buffer size too small. ...)`.

However, if we set `null_count` to be `None` in this test, we will panic at `range end index 2 out of range for slice of length 1` when calling `count_nulls(null_bit_buffer.as_ref(), offset, len)` in `new_unchecked`.

**Comment by alamb** [comment]
cc @tustvold

**Comment by tustvold** [comment]
I think I'm missing something, new_unchecked is unsafe because it trusts the caller. Making it fallible / perform validation seems counter to that, and I'm not sure I follow why it is necessary?

**Comment by HaoYang670** [comment]
> Isn't the issue in try_new?

Sorry, my mistake. I think you are right. 
We should modify `ArrayData::try_new` like that:
```rust
    pub fn try_new(
        ...
    ) -> Result<Self> {
        check_size_of_null_bit_buffer()?

        let new_self = unsafe {
            Self::new_unchecked(
                ...
            )
        };

        // As the data is not trusted, do a full validation of its contents
        new_self.validate_full()?;
        Ok(new_self)
    }
```

**Comment by HaoYang670** [comment]
I will fix this in #1707

**Comment by alamb** [comment]
🎉
